src/ros_filter.cpp

```diff
@@ -2894,32 +2965,57 @@ namespace RobotLocalization
   }
 
   template<typename T>
-  void RosFilter<T>::saveFilterState(FilterBase& filter)
-  {
-    FilterStatePtr state = FilterStatePtr(new FilterState());
-    state->state_ = Eigen::VectorXd(filter.getState());
-    state->estimateErrorCovariance_ = Eigen::MatrixXd(filter.getEstimateErrorCovariance());
-    state->lastMeasurementTime_ = filter.getLastMeasurementTime();
-    state->latestControl_ = Eigen::VectorXd(filter.getControl());
-    state->latestControlTime_ = filter.getControlTime();
-    filterStateHistory_.push_back(state);
-    RF_DEBUG("Saved state with timestamp " << std::setprecision(20) << state->lastMeasurementTime_ <<
-             " to history. " << filterStateHistory_.size() << " measurements are in the queue.\n");
-  }
-
-  template<typename T>
   bool RosFilter<T>::revertTo(const double time)
   {
     RF_DEBUG("\n----- RosFilter::revertTo -----\n");
-    RF_DEBUG("\nRequested time was " << std::setprecision(20) << time << "\n")
+    RF_DEBUG("\nRequested time was " << std::setprecision(20) << time << "\n");
 
-    // Walk back through the queue until we reach a filter state whose time stamp is less than or equal to the requested time.
-    // Since every saved state after that time will be overwritten/corrected, we can pop from the queue.
+    // Walk back through the queue until we reach a filter state whose time stamp is less than or equal to the
+    // requested time. Since every saved state after that time will be overwritten/corrected, we can pop from
+    // the queue.
     while (!filterStateHistory_.empty() && filterStateHistory_.back()->lastMeasurementTime_ > time)
     {
       filterStateHistory_.pop_back();
     }
 
+    // If we jump back in the state history, then we unfortunately need, for each differential input source, to find
+    // the *previous* filter state for that measurement (i.e., the filter state at the *previous* measurement for that
+    // sensor.
+
+    // Iterate over all the previousMeasurementStates_, since we will only have entries for the differential sensors
+    std::map<std::string, FilterStatePtr>::iterator pmsIt;
+    std::map<std::string, FilterStatePtr>::iterator pmsErased;
+    for (pmsIt = previousMeasurementStates_.begin(); pmsIt != previousMeasurementStates_.end();)
+    {
+      // Walk backwards through the history until we find the first state that resulted from a message with this topic
```

Comment:
"first state" is slightly ambiguous w.r.t. intent -- it's the first state you _find_, but it's not the first state recorded. "most recent state" would make it clear that you want the state closest to the back of the queue.